diff --git a/compiler-rt/lib/fuzzer/FuzzerDriver.cpp b/compiler-rt/lib/fuzzer/FuzzerDriver.cpp
index 306e644277c..26db440ecd7 100644
--- a/compiler-rt/lib/fuzzer/FuzzerDriver.cpp
+++ b/compiler-rt/lib/fuzzer/FuzzerDriver.cpp
@@ -737,6 +737,8 @@ int FuzzerDriver(int *argc, char ***argv, UserCallback Callback) {
       F->SetMaxInputLen(kDefaultMaxMergeLen);
     assert(Flags.merge_control_file);
     F->CrashResistantMergeInternalStep(Flags.merge_control_file);
+    if (Options.DumpCoverage)
+      TPC.DumpCoverage();
     exit(0);
   }
 
diff --git a/compiler-rt/lib/fuzzer/build.sh b/compiler-rt/lib/fuzzer/build.sh
index 504e54e3a81..cf0690271e4 100755
--- a/compiler-rt/lib/fuzzer/build.sh
+++ b/compiler-rt/lib/fuzzer/build.sh
@@ -2,7 +2,7 @@
 LIBFUZZER_SRC_DIR=$(dirname $0)
 CXX="${CXX:-clang}"
 for f in $LIBFUZZER_SRC_DIR/*.cpp; do
-  $CXX -g -O2 -fno-omit-frame-pointer -std=c++11 $f -c &
+  $CXX -stdlib=libc++ -g -O2 -fno-omit-frame-pointer -std=c++11 $f -fPIC -c &
 done
 wait
 rm -f libFuzzer.a
