diff --git a/src/afl-fuzz-queue.c b/src/afl-fuzz-queue.c
index 0d7d031..6b43a9b 100644
--- a/src/afl-fuzz-queue.c
+++ b/src/afl-fuzz-queue.c
@@ -662,26 +662,34 @@ u32 calculate_score(afl_state_t *afl, struct queue_entry *q) {
 
       switch ((u32)log2(afl->n_fuzz[q->n_fuzz_entry])) {
 
-        case 0 ... 1:
-          factor = 4;
+        case 0:
+          factor = 32;
+          break;
+
+        case 1 ... 2:
+          factor = 16;
           break;
 
-        case 2 ... 3:
-          factor = 3;
+        case 3:
+          factor = 8;
           break;
 
         case 4:
-          factor = 2;
+          factor = 4;
           break;
 
         case 5:
+          factor = 2;
           break;
 
         case 6:
-          if (!q->favored) factor = 0.8;
           break;
 
         case 7:
+          if (!q->favored) factor = 0.8;
+          break;
+
+        case 8:
           if (!q->favored) factor = 0.6;
           break;
 
