diff --git a/afl-fuzz.c b/afl-fuzz.c
index 61a3ee9..3aaeee5 100644
--- a/afl-fuzz.c
+++ b/afl-fuzz.c
@@ -4736,7 +4736,9 @@ static u32 calculate_score(struct queue_entry* q) {
      global average. Multiplier ranges from 0.1x to 3x. Fast inputs are
      less expensive to fuzz, so we're giving them more air time. */
 
-  if (q->exec_us * 0.1 > avg_exec_us) perf_score = 10;
+  if (q->exec_us * 0.01 > avg_exec_us || q->exec_us > 50000) return 1;
+  else if (q->exec_us * 0.05 > avg_exec_us || q->exec_us > 10000) return 5;
+  else if (q->exec_us * 0.1 > avg_exec_us) perf_score = 10;
   else if (q->exec_us * 0.25 > avg_exec_us) perf_score = 25;
   else if (q->exec_us * 0.5 > avg_exec_us) perf_score = 50;
   else if (q->exec_us * 0.75 > avg_exec_us) perf_score = 75;
