diff --git a/src/afl-fuzz-one.c b/src/afl-fuzz-one.c
index c04b492..b4c85bf 100644
--- a/src/afl-fuzz-one.c
+++ b/src/afl-fuzz-one.c
@@ -408,6 +408,10 @@ u8 fuzz_one_original(afl_state_t *afl) {
 
   }
 
+  /* Skip abundant entries with higher probability */
+  if (log10(afl->n_fuzz[afl->queue_cur->n_fuzz_entry]) > rand_below(afl, 10))
+    return 1;
+
   if (likely(afl->pending_favored)) {
 
     /* If we have any favored, non-fuzzed new arrivals in the queue,
